### ISSUE TITLE
Optimism Mainnet Archive crashes on startup

### ISSUE THREAD

**Issue by emlautarom1** [NethermindEth/nethermind#7880]
**Description**
When starting Nethermind using the `Optimism Mainnet Archive` configuration the node crashes after downloading the snapshot.

**Steps to Reproduce**
Run Nethermind as follows

```
dotnet run --project src/Nethermind/Nethermind.Runner/Nethermind.Runner.csproj -c Debug -- -c op-mainnet_archive
```

**Actual behavior**
Nethermind crashes with the following exception

```
2024-12-06 15:49:14.3407|ERROR|Init.Steps.EthereumStepsManager|25|Step InitializeBlockTree      failed after 200ms Nethermind.Serialization.Rlp.RlpException: VRS is 0 length when decoding Transaction
   at Nethermind.Serialization.Rlp.TxDecoders.SignatureBuilder.FromBytes(UInt64 v, ReadOnlySpan`1 rBytes, ReadOnlySpan`1 sBytes, RlpBehaviors rlpBehaviors) in /src/Nethermind/Nethermind.Serialization.Rlp/TxDecoders/SignatureBuilder.cs:line 38
   at Nethermind.Serialization.Rlp.TxDecoders.LegacyTxDecoder`1.DecodeSignature(UInt64 v, ReadOnlySpan`1 rBytes, ReadOnlySpan`1 sBytes, Signature fallbackSignature, RlpBehaviors rlpBehaviors) in /src/Nethermind/Nethermind.Serialization.Rlp/TxDecoders/LegacyTxDecoder.cs:line 34
   at Nethermind.Serialization.Rlp.TxDecoders.BaseTxDecoder`1.DecodeSignature(Transaction transaction, ValueDecoderContext& decoderContext, RlpBehaviors rlpBehaviors) in /src/Nethermind/Nethermind.Serialization.Rlp/TxDecoders/BaseTxDecoder.cs:line 171
   at Nethermind.Serialization.Rlp.TxDecoders.BaseTxDecoder`1.Decode(Transaction& transaction, Int32 txSequenceStart, ReadOnlySpan`1 transactionSequence, ValueDecoderContext& decoderContext, RlpBehaviors rlpBehaviors) in /src/Nethermind/Nethermind.Serialization.Rlp/TxDecoders/BaseTxDecoder.cs:line 73
   at Nethermind.Serialization.Rlp.TxDecoder`1.Decode(ValueDecoderContext& decoderContext, T& transaction, RlpBehaviors rlpBehaviors) in /src/Nethermind/Nethermind.Serialization.Rlp/TxDecoder.cs:line 131
   at Nethermind.Serialization.Rlp.TxDecoder`1.Decode(ValueDecoderContext& decoderContext, RlpBehaviors rlpBehaviors) in /src/Nethermind/Nethermind.Serialization.Rlp/TxDecoder.cs:line 94
   at Nethermind.Serialization.Rlp.Rlp.ValueDecoderContext.DecodeArray[T](IRlpValueDecoder`1 decoder, Boolean checkPositions, T defaultElement) in /src/Nethermind/Nethermind.Serialization.Rlp/Rlp.cs:line 1610
   at Nethermind.Serialization.Rlp.BlockBodyDecoder.DecodeUnwrapped(ValueDecoderContext& ctx, Int32 lastPosition) in /src/Nethermind/Nethermind.Serialization.Rlp/BlockBodyDecoder.cs:line 66
   at Nethermind.Serialization.Rlp.BlockDecoder.Decode(ValueDecoderContext& decoderContext, RlpBehaviors rlpBehaviors) in /src/Nethermind/Nethermind.Serialization.Rlp/BlockDecoder.cs:line 70
   at Nethermind.Serialization.Rlp.KeyValueStoreRlpExtensions.Get[TCacheKey,TItem](IReadOnlyKeyValueStore db, TCacheKey cacheKey, ReadOnlySpan`1 key, IRlpStreamDecoder`1 decoder, ClockCache`2 cache, RlpBehaviors rlpBehaviors, Boolean shouldCache) in /src/Nethermind/Nethermind.Serialization.Rlp/KeyValueStoreRlpExtensions.cs:line 66
   at Nethermind.Serialization.Rlp.KeyValueStoreRlpExtensions.Get[TItem](IReadOnlyKeyValueStore db, ValueHash256 key, IRlpStreamDecoder`1 decoder, ClockCache`2 cache, RlpBehaviors rlpBehaviors, Boolean shouldCache) in /src/Nethermind/Nethermind.Serialization.Rlp/KeyValueStoreRlpExtensions.cs:line 27
   at Nethermind.Blockchain.Blocks.BlockStore.Get(Int64 blockNumber, Hash256 blockHash, RlpBehaviors rlpBehaviors, Boolean shouldCache) in /src/Nethermind/Nethermind.Blockchain/Blocks/BlockStore.cs:line 71
   at Nethermind.Blockchain.BlockTree.FindBlock(Hash256 blockHash, BlockTreeLookupOptions options, Nullable`1 blockNumber) in /src/Nethermind/Nethermind.Blockchain/BlockTree.cs:line 1261
   at Nethermind.Blockchain.BlockTree.BodyExists(Int64 blockNumber, Boolean findBeacon) in /src/Nethermind/Nethermind.Blockchain/BlockTree.Initializer.cs:line 93
   at Nethermind.Blockchain.BlockTree.BinarySearchBlockNumber(Int64 left, Int64 right, Func`3 isBlockFound, BinarySearchDirection direction, Boolean findBeacon) in /src/Nethermind/Nethermind.Blockchain/BlockTree.Initializer.cs:line 276
   at Nethermind.Blockchain.BlockTree.LoadBeaconBestKnown() in /src/Nethermind/Nethermind.Blockchain/BlockTree.Initializer.cs:line 203
   at Nethermind.Blockchain.BlockTree.RecalculateTreeLevels() in /src/Nethermind/Nethermind.Blockchain/BlockTree.Initializer.cs:line 22
   at Nethermind.Blockchain.BlockTree..ctor(IBlockStore blockStore, IHeaderStore headerDb, IDb blockInfoDb, IDb metadataDb, IBadBlockStore badBlockStore, IChainLevelInfoRepository chainLevelInfoRepository, ISpecProvider specProvider, IBloomStorage bloomStorage, ISyncConfig syncConfig, ILogManager logManager) in /src/Nethermind/Nethermind.Blockchain/BlockTree.cs:line 147
   at Nethermind.Init.Steps.InitializeBlockTree.Execute(CancellationToken cancellationToken) in /src/Nethermind/Nethermind.Init/Steps/InitializeBlockTree.cs:line 57
   at Nethermind.Init.Steps.EthereumStepsManager.ExecuteStep(IStep step, StepInfo stepInfo, CancellationToken cancellationToken) in /src/Nethermind/Nethermind.Init/Steps/EthereumStepsManager.cs:line 153
2024-12-06 15:49:14.3577|ERROR|Program|25|A critical error has occurred Nethermind.Serialization.Rlp.RlpException: VRS is 0 length when decoding Transaction
   at Nethermind.Serialization.Rlp.TxDecoders.SignatureBuilder.FromBytes(UInt64 v, ReadOnlySpan`1 rBytes, ReadOnlySpan`1 sBytes, RlpBehaviors rlpBehaviors) in /src/Nethermind/Nethermind.Serialization.Rlp/TxDecoders/SignatureBuilder.cs:line 38
   at Nethermind.Serialization.Rlp.TxDecoders.LegacyTxDecoder`1.DecodeSignature(UInt64 v, ReadOnlySpan`1 rBytes, ReadOnlySpan`1 sBytes, Signature fallbackSignature, RlpBehaviors rlpBehaviors) in /src/Nethermind/Nethermind.Serialization.Rlp/TxDecoders/LegacyTxDecoder.cs:line 34
   at Nethermind.Serialization.Rlp.TxDecoders.BaseTxDecoder`1.DecodeSignature(Transaction transaction, ValueDecoderContext& decoderContext, RlpBehaviors rlpBehaviors) in /src/Nethermind/Nethermind.Serialization.Rlp/TxDecoders/BaseTxDecoder.cs:line 171
   at Nethermind.Serialization.Rlp.TxDecoders.BaseTxDecoder`1.Decode(Transaction& transaction, Int32 txSequenceStart, ReadOnlySpan`1 transactionSequence, ValueDecoderContext& decoderContext, RlpBehaviors rlpBehaviors) in /src/Nethermind/Nethermind.Serialization.Rlp/TxDecoders/BaseTxDecoder.cs:line 73
   at Nethermind.Serialization.Rlp.TxDecoder`1.Decode(ValueDecoderContext& decoderContext, T& transaction, RlpBehaviors rlpBehaviors) in /src/Nethermind/Nethermind.Serialization.Rlp/TxDecoder.cs:line 131
   at Nethermind.Serialization.Rlp.TxDecoder`1.Decode(ValueDecoderContext& decoderContext, RlpBehaviors rlpBehaviors) in /src/Nethermind/Nethermind.Serialization.Rlp/TxDecoder.cs:line 94
   at Nethermind.Serialization.Rlp.Rlp.ValueDecoderContext.DecodeArray[T](IRlpValueDecoder`1 decoder, Boolean checkPositions, T defaultElement) in /src/Nethermind/Nethermind.Serialization.Rlp/Rlp.cs:line 1610
   at Nethermind.Serialization.Rlp.BlockBodyDecoder.DecodeUnwrapped(ValueDecoderContext& ctx, Int32 lastPosition) in /src/Nethermind/Nethermind.Serialization.Rlp/BlockBodyDecoder.cs:line 66
   at Nethermind.Serialization.Rlp.BlockDecoder.Decode(ValueDecoderContext& decoderContext, RlpBehaviors rlpBehaviors) in /src/Nethermind/Nethermind.Serialization.Rlp/BlockDecoder.cs:line 70
   at Nethermind.Serialization.Rlp.KeyValueStoreRlpExtensions.Get[TCacheKey,TItem](IReadOnlyKeyValueStore db, TCacheKey cacheKey, ReadOnlySpan`1 key, IRlpStreamDecoder`1 decoder, ClockCache`2 cache, RlpBehaviors rlpBehaviors, Boolean shouldCache) in /src/Nethermind/Nethermind.Serialization.Rlp/KeyValueStoreRlpExtensions.cs:line 66
   at Nethermind.Serialization.Rlp.KeyValueStoreRlpExtensions.Get[TItem](IReadOnlyKeyValueStore db, ValueHash256 key, IRlpStreamDecoder`1 decoder, ClockCache`2 cache, RlpBehaviors rlpBehaviors, Boolean shouldCache) in /src/Nethermind/Nethermind.Serialization.Rlp/KeyValueStoreRlpExtensions.cs:line 27
   at Nethermind.Blockchain.Blocks.BlockStore.Get(Int64 blockNumber, Hash256 blockHash, RlpBehaviors rlpBehaviors, Boolean shouldCache) in /src/Nethermind/Nethermind.Blockchain/Blocks/BlockStore.cs:line 71
   at Nethermind.Blockchain.BlockTree.FindBlock(Hash256 blockHash, BlockTreeLookupOptions options, Nullable`1 blockNumber) in /src/Nethermind/Nethermind.Blockchain/BlockTree.cs:line 1261
   at Nethermind.Blockchain.BlockTree.BodyExists(Int64 blockNumber, Boolean findBeacon) in /src/Nethermind/Nethermind.Blockchain/BlockTree.Initializer.cs:line 93
   at Nethermind.Blockchain.BlockTree.BinarySearchBlockNumber(Int64 left, Int64 right, Func`3 isBlockFound, BinarySearchDirection direction, Boolean findBeacon) in /src/Nethermind/Nethermind.Blockchain/BlockTree.Initializer.cs:line 276
   at Nethermind.Blockchain.BlockTree.LoadBeaconBestKnown() in /src/Nethermind/Nethermind.Blockchain/BlockTree.Initializer.cs:line 203
   at Nethermind.Blockchain.BlockTree.RecalculateTreeLevels() in /src/Nethermind/Nethermind.Blockchain/BlockTree.Initializer.cs:line 22
   at Nethermind.Blockchain.BlockTree..ctor(IBlockStore blockStore, IHeaderStore headerDb, IDb blockInfoDb, IDb metadataDb, IBadBlockStore badBlockStore, IChainLevelInfoRepository chainLevelInfoRepository, ISpecProvider specProvider, IBloomStorage bloomStorage, ISyncConfig syncConfig, ILogManager logManager) in /src/Nethermind/Nethermind.Blockchain/BlockTree.cs:line 147
   at Nethermind.Init.Steps.InitializeBlockTree.Execute(CancellationToken cancellationToken) in /src/Nethermind/Nethermind.Init/Steps/InitializeBlockTree.cs:line 57
   at Nethermind.Init.Steps.EthereumStepsManager.ExecuteStep(IStep step, StepInfo stepInfo, CancellationToken cancellationToken) in /src/Nethermind/Nethermind.Init/Steps/EthereumStepsManager.cs:line 153
   at Nethermind.Init.Steps.EthereumStepsManager.ReviewFailedAndThrow() in /src/Nethermind/Nethermind.Init/Steps/EthereumStepsManager.cs:line 210
   at Nethermind.Init.Steps.EthereumStepsManager.InitializeAll(CancellationToken cancellationToken) in /src/Nethermind/Nethermind.Init/Steps/EthereumStepsManager.cs:line 95
   at Nethermind.Runner.Ethereum.EthereumRunner.Start(CancellationToken cancellationToken) in /src/Nethermind/Nethermind.Runner/Ethereum/EthereumRunner.cs:line 30
   at Program.<>c__DisplayClass0_0.<<<Main>$>g__RunAsync|2>d.MoveNext() in /src/Nethermind/Nethermind.Runner/Program.cs:line 208
2024-12-06 15:49:14.3577|INFO|Program|25|Nethermind is shutting down... Please wait until all activities are stopped.
```

**Expected behavior**
Nethermind should start normally.

**Desktop (please complete the following information):**
Please provide the following information regarding your setup:
 - Operating System: Fedora Workstation 41
 - Version: `1.30.0-rc`
 - Installation Method: From source
 - Consensus Client: irrelevant

**Comment by emlautarom1** [comment]
I was able to narrow down the issue to block [105235062](https://optimism.blockscout.com/block/0x21a168dfa5e727926063a28ba16fd5ee84c814e847c81a699c7a0ea551e4ca50?tab=index), specifically when decoding its ownly transaction [0xd4c3dab63aaf00a4e3811ffd2e72c2de73e7ff7507e7b5f8e79a4b700f0cf227](https://optimism.blockscout.com/tx/0xd4c3dab63aaf00a4e3811ffd2e72c2de73e7ff7507e7b5f8e79a4b700f0cf227?tab=index)

If the data in the snapshot is correct and I didn't mess up, the RLP for the above mentioned block is the following:

```
0xf90430f9025da0d905ae955c0dafb7cdd1900e08289f6c3e939388c7b810e7de4eb441aabf9d59a01dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347940000000000000000000000000000000000000000a04adc41ce7ddde6398c47f5e2fbe3f31754268f12e4f14b2a575cf9d5b5363bd3a0353169389d80e656f9da5866eb1b78d3397a3d947c71753194bbd1a8a6108076a01e415689e75e50719c2d9714995cd9615edabffb3390ad908803ca21d2df96fbb9010000000000000000000000000000000000000000000000000000000000001000000000000000000080000000000000000100000000000000000000004000000000008000000024000000000008000000000000000000000000000000000000002000000000020000400000000000020800200000200000400000000018000000004000000000000000000000000000000001800000000000000000000000000000000080040000000000000000000000000000a0000000000000000000000000400000000200000000020000000000000000000212000000000000000000002000000000080000000000000000000000000000000000000800000000000800000102840645c27683e4e1c083020e8d84647f587db861d98301090a846765746889676f312e31352e3133856c696e75780000000000006579f2174bcc83cf381ad2f867e287448bc3472d40fe92a35e77a6922cbd033b0fff6a2b60929856dfd7993bac8626cd57a77daf86247a47d4d8e90d5f653e9e01a00000000000000000000000000000000000000000000000000000000000000000880000000000000000f901ccf901c9830571188083030d4094420000000000000000000000000000000000000780b901a4cbd4ece9000000000000000000000000420000000000000000000000000000000000001000000000000000000000000099c9fc46f92e8a1c0dec1b1747d010903e884be10000000000000000000000000000000000000000000000000000000000000080000000000000000000000000000000000000000000000000000000000005711800000000000000000000000000000000000000000000000000000000000000e4662a633a000000000000000000000000dac17f958d2ee523a2206206994597c13d831ec700000000000000000000000094b008aa00579c1307b0ef2c499ad98a8ce58e58000000000000000000000000117274dde02bc94006185af87d78beab28ceae06000000000000000000000000117274dde02bc94006185af87d78beab28ceae06000000000000000000000000000000000000000000000000000000000c3d8b8000000000000000000000000000000000000000000000000000000000000000c0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000808080c0
```

We can then create a test that reproduces the issue:

```c#
[Test]
public void Problematic_Archive_Block()
{
    TxDecoder.Instance.RegisterDecoder(new OptimismTxDecoder<Transaction>());

    var hexBytes =
        "f90430f9025da0d905ae955c0dafb7cdd1900e08289f6c3e939388c7b810e7de4eb441aabf9d59a01dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347940000000000000000000000000000000000000000a04adc41ce7ddde6398c47f5e2fbe3f31754268f12e4f14b2a575cf9d5b5363bd3a0353169389d80e656f9da5866eb1b78d3397a3d947c71753194bbd1a8a6108076a01e415689e75e50719c2d9714995cd9615edabffb3390ad908803ca21d2df96fbb9010000000000000000000000000000000000000000000000000000000000001000000000000000000080000000000000000100000000000000000000004000000000008000000024000000000008000000000000000000000000000000000000002000000000020000400000000000020800200000200000400000000018000000004000000000000000000000000000000001800000000000000000000000000000000080040000000000000000000000000000a0000000000000000000000000400000000200000000020000000000000000000212000000000000000000002000000000080000000000000000000000000000000000000800000000000800000102840645c27683e4e1c083020e8d84647f587db861d98301090a846765746889676f312e31352e3133856c696e75780000000000006579f2174bcc83cf381ad2f867e287448bc3472d40fe92a35e77a6922cbd033b0fff6a2b60929856dfd7993bac8626cd57a77daf86247a47d4d8e90d5f653e9e01a00000000000000000000000000000000000000000000000000000000000000000880000000000000000f901ccf901c9830571188083030d4094420000000000000000000000000000000000000780b901a4cbd4ece9000000000000000000000000420000000000000000000000000000000000001000000000000000000000000099c9fc46f92e8a1c0dec1b1747d010903e884be10000000000000000000000000000000000000000000000000000000000000080000000000000000000000000000000000000000000000000000000000005711800000000000000000000000000000000000000000000000000000000000000e4662a633a000000000000000000000000dac17f958d2ee523a2206206994597c13d831ec700000000000000000000000094b008aa00579c1307b0ef2c499ad98a8ce58e58000000000000000000000000117274dde02bc94006185af87d78beab28ceae06000000000000000000000000117274dde02bc94006185af87d78beab28ceae06000000000000000000000000000000000000000000000000000000000c3d8b8000000000000000000000000000000000000000000000000000000000000000c0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000808080c0";
    var bytes = Bytes.FromHexString(hexBytes);
    var context = bytes.AsRlpValueContext();

    var decoder = new BlockDecoder();
    var block = decoder.Decode(ref context,  RlpBehaviors.None | RlpBehaviors.AllowExtraBytes);
}
```

Unfortunately I don't have a way to confirm that the RLP listed above is correct. We could get it from a trusted node through one of the following methods:

```json
{
  "jsonrpc": "2.0",
  "method": "debug_getRawTransaction",
  "params": ["0xd4c3dab63aaf00a4e3811ffd2e72c2de73e7ff7507e7b5f8e79a4b700f0cf227"],
  "id": 1
}
```

or 

```json
{
  "jsonrpc": "2.0",
  "method": "debug_getRawBlock",
  "params": ["0x645C276"],
  "id": 1
}
```

**Comment by emlautarom1** [comment]
The correct RLP for the problematic transaction is the following:

```
f901c9830571188083030d4094420000000000000000000000000000000000000780b901a4cbd4ece9000000000000000000000000420000000000000000000000000000000000001000000000000000000000000099c9fc46f92e8a1c0dec1b1747d010903e884be10000000000000000000000000000000000000000000000000000000000000080000000000000000000000000000000000000000000000000000000000005711800000000000000000000000000000000000000000000000000000000000000e4662a633a000000000000000000000000dac17f958d2ee523a2206206994597c13d831ec700000000000000000000000094b008aa00579c1307b0ef2c499ad98a8ce58e58000000000000000000000000117274dde02bc94006185af87d78beab28ceae06000000000000000000000000117274dde02bc94006185af87d78beab28ceae06000000000000000000000000000000000000000000000000000000000c3d8b8000000000000000000000000000000000000000000000000000000000000000c0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000808080
```

The following test reproduces the error:

```c#
[Test]
public void Problematic_Archive_Block()
{
    _decoder.RegisterDecoder(new OptimismTxDecoder<Transaction>());

    var hexBytes =
        "f901c9830571188083030d4094420000000000000000000000000000000000000780b901a4cbd4ece9000000000000000000000000420000000000000000000000000000000000001000000000000000000000000099c9fc46f92e8a1c0dec1b1747d010903e884be10000000000000000000000000000000000000000000000000000000000000080000000000000000000000000000000000000000000000000000000000005711800000000000000000000000000000000000000000000000000000000000000e4662a633a000000000000000000000000dac17f958d2ee523a2206206994597c13d831ec700000000000000000000000094b008aa00579c1307b0ef2c499ad98a8ce58e58000000000000000000000000117274dde02bc94006185af87d78beab28ceae06000000000000000000000000117274dde02bc94006185af87d78beab28ceae06000000000000000000000000000000000000000000000000000000000c3d8b8000000000000000000000000000000000000000000000000000000000000000c0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000808080";
    var bytes = Bytes.FromHexString(hexBytes);
    var context = bytes.AsRlpValueContext();

    var decoder = TxDecoder.Instance;
    var transaction = decoder.Decode(ref context);

    transaction.Should().NotBeNull();
}
```

**Comment by emlautarom1** [comment]
Fixed in #7881 and #7887.

---

The issue seems to be that pre-bedrock we should not validate signatures in Legacy transactions. It's not clear to me why we didn't face any issues earlier, or if we did, why they went unnoticed.

I did revert the `TxDecoder` changes in #7334 to see if this was a regression but that does not seem to be the case: the transaction gets rejected due to an invalid signature too.

In op-geth (tag `v1.101411.2`), no validation is performed at all during RLP decoding, so the transaction gets accepted. During JSON deserialization, the signature fields are validated only when they're non-zero:

https://github.com/ethereum-optimism/op-geth/blob/3dd9b0274bae3d3d2c80ef517563a360108e8cf6/core/types/transaction_marshalling.go#L248-L251
